### PR TITLE
[TS migration] Migrate VideoPlayer component files to TypeScript

### PR DESCRIPTION
**Pull request by smelaa**
### Fixed Issues

$https://github.com/Expensify/App/issues/36722
PROPOSAL:

### Tests

1. Upload few videos to one chat
2. Play and pause one of them
3. Change volume
4. Turn off and on volume
5. Change video speed
6. Enter full-screen mode and repeat the steps
7. Move to another video
8. Verify that no errors appear in the JS console
- [x] Verify that no errors appear in the JS console
### Offline tests

N/A

### QA Steps

Same as tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[android.webm](https://github.com/Expensify/App/assets/115273836/2b09b1ee-8f74-47fe-b07e-76858b102230)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[android_web.webm](https://github.com/Expensify/App/assets/115273836/ee890564-43f8-44f0-82d2-3717612083b8)

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/115273836/673bff13-7dad-4d70-ade1-5949c09a9bc7

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/115273836/97da3ab0-fff8-4da6-a89b-b692af616104

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/115273836/21b4d8ab-735e-4968-abd2-0caa7da4ef32

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/115273836/51cbd441-2330-4094-b4d6-227cc5e6ffcf

</details>

### CODE REVIEW

**Comment by fabioh8010** [comment]
Hey @smelaa, why is this PR on hold?

**Review by war-in — APPROVED** [review]

File: src/components/VideoPlayer/BaseVideoPlayer.tsx [hunk 1]
```diff
@@ -99,34 +103,35 @@ function BaseVideoPlayer({
     );
 
     const handlePlaybackStatusUpdate = useCallback(
-        (e) => {
-            const isVideoPlaying = e.isPlaying || false;
-            const currentDuration = e.durationMillis || videoDuration * 1000;
-            const currentPositon = e.positionMillis || 0;
+        (status: AVPlaybackStatus) => {
+            const isVideoPlaying = 'isPlaying' in status ? status.isPlaying : false;
```
Comment (blazejkustra) [inline src/components/VideoPlayer/BaseVideoPlayer.tsx:107]: Let's add an early return here, this way the logic can be much simpler. Wdyt @smelaa?

File: src/components/VideoPlayer/IconButton.tsx [hunk 2]
```diff
@@ -1,42 +1,23 @@
...
+    fill?: string;
+    tooltipText?: string;
+    style?: StyleProp<ViewStyle>;
+    hoverStyle?: StyleProp<ViewStyle>;
+    small?: boolean;
+    shouldForceRenderingTooltipBelow?: boolean;
 };
 
-function IconButton({src, fill, onPress, style, hoverStyle, tooltipText, small, shouldForceRenderingTooltipBelow}) {
+function IconButton({src, fill = 'white', onPress = () => {}, style = {}, hoverStyle = {}, tooltipText = '', small = false, shouldForceRenderingTooltipBelow = false}: IconButtonProps) {
```
Comment (blazejkustra) [inline src/components/VideoPlayer/IconButton.tsx:20]: Let's remove empty objects for styles, and maybe onPress too?

(styles work for falsy values too, and onPress can be undefined and we can add optional chaining like: `onPress?.()`)

File: src/components/VideoPlayer/index.native.tsx [hunk 3]
```diff
@@ -1,23 +1,20 @@
...
-function VideoPlayer({videoControlsStyle, ...props}, ref) {
+function VideoPlayer({videoControlsStyle, ...props}: VideoPlayerProps) {
     return (
         <BaseVideoPlayer
             // eslint-disable-next-line react/jsx-props-no-spreading
             {...props}
             isVideoHovered
             shouldUseSharedVideoElement={false}
             videoControlsStyle={[{bottom: CONST.VIDEO_PLAYER.CONTROLS_POSITION.NATIVE}, videoControlsStyle]}
-            ref={ref}
```
Comment (blazejkustra) [inline src/components/VideoPlayer/index.native.tsx:14]: Why is this removed? Was it unused? Then why is forwardRef still used in the file? 🤔

File: src/components/VideoPlayer/index.tsx [hunk 4]
```diff
@@ -1,19 +1,16 @@
...
-import {videoPlayerDefaultProps, videoPlayerPropTypes} from './propTypes';
+import type VideoPlayerProps from './types';
 
-function VideoPlayer(props, ref) {
+function VideoPlayer(props: VideoPlayerProps) {
     return (
         <BaseVideoPlayer
             // eslint-disable-next-line react/jsx-props-no-spreading
             {...props}
-            ref={ref}
```
Comment (blazejkustra) [inline src/components/VideoPlayer/index.tsx:10]: Same here

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/components/VideoPlayerPreview/index.tsx [hunk 5]
```diff
@@ -82,7 +82,6 @@ function VideoPlayerPreview({videoUrl, thumbnailUrl, fileName, videoDimensions,
                 <>
                     <VideoPlayer
                         url={videoUrl}
-                        onOpenInModalButtonPress={onShowModalPress}
```
[the remark this answers is not in the document]
Comment (smelaa) [reply src/components/VideoPlayerPreview/index.tsx:85]: Yes, `onOpenInModalButtonPress` is not a VideoPlayer prop. I think this part of the code was missed when the prop was removed.

File: src/components/VideoPlayer/index.native.tsx (hunk 3 above)
Comment (smelaa) [reply src/components/VideoPlayer/index.native.tsx:14]: It's unused. I also removed forwardRef.

File: src/components/VideoPlayer/BaseVideoPlayer.tsx [hunk 6]
```diff
@@ -144,11 +160,16 @@ function BaseVideoPlayer({
...
-        // Update states after binding
-        currentVideoPlayerRef.current.getStatusAsync().then((status) => {
+        const currentVideoPlayer = currentVideoPlayerRef.current;
+        if (!currentVideoPlayer) {
+            return;
+        }
+        currentVideoPlayer._onPlaybackStatusUpdate = handlePlaybackStatusUpdate;
+        if ('_onFullscreenUpdate' in currentVideoPlayer) {
+            currentVideoPlayer._onFullscreenUpdate = handleFullscreenUpdate;
+        }
```
Comment (blazejkustra) [inline src/components/VideoPlayer/BaseVideoPlayer.tsx:170]: This changes a logic a bit, let's make sure the function is bound properly

File: src/components/VideoPlayer/BaseVideoPlayer.tsx [hunk 7]
```diff
@@ -183,26 +204,25 @@ function BaseVideoPlayer({
         if (shouldUseSharedVideoElement || url !== currentlyPlayingURL) {
             return;
         }
-        shareVideoPlayerElements(videoPlayerRef.current, videoPlayerElementParentRef.current, videoPlayerElementRef.current, isUploading);
-    }, [currentlyPlayingURL, shouldUseSharedVideoElement, shareVideoPlayerElements, updateSharedElements, url, isUploading]);
+        shareVideoPlayerElements(videoPlayerRef.current, videoPlayerElementParentRef.current as View | null, videoPlayerElementRef.current as View | null, isUploading);
```
Comment (blazejkustra) [inline src/components/VideoPlayer/BaseVideoPlayer.tsx:207]: Shouldn't be `shareVideoPlayerElements` adjusted instead of assertion?

**Review by blazejkustra — APPROVED** [review]

**Review by Skalakid — APPROVED** [review]

File: src/components/VideoPlayer/BaseVideoPlayer.tsx [hunk 8]
```diff
@@ -14,38 +16,37 @@ import addEncryptedAuthTokenToURL from '@libs/addEncryptedAuthTokenToURL';
...
-    shouldUseSharedVideoElement,
-    shouldUseSmallVideoControls,
-    shouldShowVideoControls,
-    onPlaybackStatusUpdate,
-    onFullscreenUpdate,
+    shouldUseSharedVideoElement = false,
+    shouldUseSmallVideoControls = false,
+    shouldShowVideoControls = true,
+    onPlaybackStatusUpdate = () => {},
+    onFullscreenUpdate = () => {},
```
Comment (kubabutkiewicz) [inline src/components/VideoPlayer/BaseVideoPlayer.tsx:40]: I think the default values for a callbacks are not needed

Comment (smelaa) [reply src/components/VideoPlayer/BaseVideoPlayer.tsx:40]: Are you sure? Without them we might invoke object which is possibly undefined. Those props are optional as some BaseVideoPlayer calls does not provide them.

Comment (kubabutkiewicz) [reply src/components/VideoPlayer/BaseVideoPlayer.tsx:40]: yeah, and when calling those functions you can use optional chaining eg. `onPlaybackStatusUpdate.?()` then if the function is undefined nothing will happen 😄

Comment (smelaa) [reply src/components/VideoPlayer/BaseVideoPlayer.tsx]: Oh yes, that's right. I have removed default callbacks then.

**Review by Skalakid — APPROVED** [review]

**Review by kubabutkiewicz — APPROVED** [review]

**Comment by kubabutkiewicz** [comment]
@jjcoffee kind bump 😄

**Comment by jjcoffee** [comment]
@kubabutkiewicz Apologies I had higher priority items to review today, I will get back to this on Monday! :pray:

**Comment by jjcoffee** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by jjcoffee** [comment]
@smelaa I'm getting a crash when navigating between videos:

image


https://github.com/Expensify/App/assets/27287420/ce363fc5-1b5e-4ac7-8edb-a303ad154d75

**Comment by jjcoffee** [comment]
`BUG:` similar to above, but there's a crash when the video ends too (if open in the attachment modal), same error in the console, but also this additional one:

image

**Comment by fabioh8010** [comment]
@smelaa kind bump 🙂

**Comment by filip-solecki** [comment]
@jjcoffee Bugs fixed!

**Review by fabioh8010 — APPROVED** [review]

**Comment by blazejkustra** [comment]
@jjcoffee Conflicts resolved 😄

**Review by jjcoffee — APPROVED** [review]
LGTM!

**Comment by blazejkustra** [comment]
@hayata-suenaga All yours! 😄

**Review by hayata-suenaga — APPROVED** [review]
Thank you, @smelaa, for your work! 🎉
